File: backend/src/api/dtos/auth_dto.py
import uuid

from pydantic import BaseModel, EmailStr
from typing import Optional

from .user_dto import UserResponse # Use our existing User DTO as UserResponse

# ==================================
# Registration DTOs
# ==================================

class SendEmailVerificationRequest(BaseModel):
    email: EmailStr

class VerifyEmailCodeRequest(BaseModel):
    email: EmailStr
    otp: str # Assuming OTP is a 6-digit string

class CompleteRegistrationRequest(BaseModel):
    verification_token: str # A temporary token received after OTP verification
    username: str
    password: str # This will be hashed in the service layer

class ResendOTPRequest(BaseModel):
    email: EmailStr

# ==================================
# Login DTOs
# ==================================

class UserLoginRequest(BaseModel):
    identifier: str # Can be email or username
    password: str

# ==================================
# Google Setup DTOs
# ==================================

class CompleteGoogleSetupRequest(BaseModel):
    setup_token: str # A temporary token from Google OAuth process
    username: str

# ==================================
# Token Management DTOs
# ==================================

class RefreshRequest(BaseModel):
    refresh_token: str

class LogoutRequest(BaseModel):
    access_token: str
    refresh_token: str

# ==================================
# Response DTOs
# ==================================

class AuthResponse(BaseModel):
    user: UserResponse # Uses the User DTO from user_dto.py
    access_token: str
    refresh_token: str
    token_type: str = "bearer" # Default token type

class RefreshResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer" # Default token type

# DTO for the data encoded in the JWT
class TokenPayload(BaseModel):
    sub: Optional[uuid.UUID] = None

File: apps/api-gateway/internal/dto/auth_dto.go
package dto

type SendEmailVerificationRequest struct {
	Email string `json:"email" binding:"required,email"`
}

type VerifyEmailCodeRequest struct {
	Email string `json:"email" binding:"required,email"`
	OTP   string `json:"otp" binding:"required,len=6"`
}

type CompleteRegistrationRequest struct {
	VerificationToken string `json:"verification_token" binding:"required"`
	Username          string `json:"username" binding:"required,min=3,max=20"`
	Password          string `json:"password" binding:"required,min=6"`
}

type ResendOTPRequest struct {
	Email string `json:"email" binding:"required,email"`
}

// Login
type UserLoginRequest struct {
	Identifier string `json:"identifier" binding:"required"`
	Password   string `json:"password" binding:"required"`
}

type CompleteGoogleSetupRequest struct {
	SetupToken string `json:"setup_token" binding:"required"`
	Username   string `json:"username" binding:"required,min=3,max=20"`
}
type RefreshRequest struct {
	RefreshToken string `json:"refresh_token" binding:"required"`
}

type LogoutRequest struct {
	AccessToken  string `json:"access_token" binding:"required"`
	RefreshToken string `json:"refresh_token" binding:"required"`
}

// AuthResponse is returned on successful login or registration.
type AuthResponse struct {
	User         *UserResponse `json:"user"`
	AccessToken  string        `json:"access_token"`
	RefreshToken string        `json:"refresh_token"`
}

type RefreshResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
}
